File: Ch. 06/Matched Pairs (Pilot)/code/optimal_matching_from_pilot.R
# =============================================================================
# Optimal Matched-Pairs Randomization Using Pilot Study
# =============================================================================
# Implements optimal matched-pairs randomization based on Bai (2022) framework
# for minimizing mean-squared error. Uses pilot study regression to estimate
# expected outcomes g_i = E[Y_i(1) + Y_i(0) | X_i] for main sample, then creates
# matched pairs by sorting on g_i and pairing adjacent units.
#
# Steps:
#   1. Load pilot data (with Treatment and Outcome from previous randomization)
#   2. Run regression on pilot to estimate relationship between covariates and outcomes
#   3. For main sample: predict outcomes under D=0 and D=1
#   4. Calculate g_hat = g_0 + g_1 for each unit
#   5. Sort main sample by g_hat
#   6. Create matched pairs from sorted data (pair adjacent units)
#   7. Within each pair, randomly assign treatment
#
# Section 6.3.5.1: Efficient Matching Minimizing Mean-Squared Error

# --- Setup -------------------------------------------------------------------
rm(list = ls())

# Load required libraries
library(tidyverse)
library(haven)

# Automatically set working directory to the folder containing this script.
# Supports RStudio (Source/Knit), source(), and Rscript from the command line.
if (requireNamespace("rstudioapi", quietly = TRUE) && rstudioapi::isAvailable()) {
  setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
} else if (length(sys.frames()) > 0 && !is.null(sys.frame(1)$ofile)) {
  setwd(dirname(sys.frame(1)$ofile))
} else if (!is.null(commandArgs(trailingOnly = FALSE))) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    setwd(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
}

# Define paths relative to script location
output_dir <- file.path("..", "output")
data_dir <- file.path("..", "data")
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)


# --- Parameters --------------------------------------------------------------
# TODO: Set random seed for reproducibility
random_seed <- 42

# TODO: Specify control variables to use in g_i estimation
# These should be the covariates available in your dataset
# IMPORTANT: Must match the control variables used in the pilot regression
control_vars <- c("female", "race_w", "birthweight", "std_ncog_pre", "year")


# --- Initialize random state -------------------------------------------------
set.seed(random_seed)


# --- Load pilot data and estimate g_i model ---------------------------------
# TODO: Specify the pilot data file
# This file should contain:
#   - Treatment variable (0/1 or binary)
#   - Outcome variable (continuous)
#   - All control variables specified in control_vars
pilot_data <- read_dta(file.path(output_dir, "pilot_sample_with_treatment_and_outcome.dta"))

# Prepare regression data (drop missing values in controls)
pilot_reg <- pilot_data %>%
  dplyr::select(Outcome, Treatment, all_of(control_vars)) %>%
  drop_na()

# TODO: Modify regression specification if desired
# You can add interaction effects, polynomials, or other transformations
# Example: pilot_formula <- as.formula("Outcome ~ Treatment + female + race_w + female:race_w + I(birthweight^2)")
# Run regression: Outcome ~ Treatment + Controls
pilot_formula <- as.formula(paste("Outcome ~ Treatment +", paste(control_vars, collapse = " + ")))
pilot_model <- lm(pilot_formula, data = pilot_reg)

# Extract model statistics
r2_pilot <- summary(pilot_model)$r.squared
r2_adj_pilot <- summary(pilot_model)$adj.r.squared
coef_treatment <- coef(pilot_model)["Treatment"]
pval_treatment <- summary(pilot_model)$coefficients["Treatment", "Pr(>|t|)"]


# --- Load main sample and estimate g_0 and g_1 ------------------------------
# TODO: Specify the main sample data file
# This file should contain:
#   - All control variables specified in control_vars
#   - NO Treatment variable (will be assigned by this script)
main_data <- read_dta(file.path(output_dir, "main_sample.dta"))

# Keep only complete cases for prediction
main_complete <- main_data %>%
  dplyr::select(all_of(control_vars)) %>%
  drop_na() %>%
  mutate(original_index = row_number())

n_complete <- nrow(main_complete)

# TODO: If you modified the regression specification above,
# you must apply the SAME transformations here for prediction
# Example: If you added interaction terms or polynomials in pilot_formula,
#          you must also create those features in main_d0 and main_d1 below
# Predict under D=0 (Treatment=0)
main_d0 <- main_complete %>%
  mutate(Treatment = 0)
main_d0$g_0 <- predict(pilot_model, newdata = main_d0)

# Predict under D=1 (Treatment=1)
main_d1 <- main_complete %>%
  mutate(Treatment = 1)
main_d1$g_1 <- predict(pilot_model, newdata = main_d1)

# Add predictions to main_complete
main_complete$g_0 <- main_d0$g_0
main_complete$g_1 <- main_d1$g_1

# Calculate g_hat = g_0 + g_1
main_complete <- main_complete %>%
  mutate(g_hat = g_0 + g_1)

# Calculate summary statistics
mean_g_hat <- mean(main_complete$g_hat)
sd_g_hat <- sd(main_complete$g_hat)
min_g_hat <- min(main_complete$g_hat)
max_g_hat <- max(main_complete$g_hat)


# --- Sort by g_hat and create matched pairs ---------------------------------
# Sort by g_hat (increasing order)
sorted_data <- main_complete %>%
  arrange(g_hat) %>%
  mutate(sorted_index = row_number())

# Create pairs from adjacent units in sorted order
n_units <- nrow(sorted_data)
n_pairs <- floor(n_units / 2)
n_unpaired <- n_units %% 2

# Initialize treatment and pair ID columns
sorted_data$Treatment_Final <- -1
sorted_data$Pair_ID <- -1

# Create pairs from adjacent units
pair_id <- 0
for (i in seq(1, 2 * n_pairs, by = 2)) {
  if (i + 1 <= nrow(sorted_data)) {
    # Randomly assign treatment within pair
    if (runif(1) < 0.5) {
      sorted_data$Treatment_Final[i] <- 1
      sorted_data$Treatment_Final[i + 1] <- 0
    } else {
      sorted_data$Treatment_Final[i] <- 0
      sorted_data$Treatment_Final[i + 1] <- 1
    }

    # Assign same pair ID
    sorted_data$Pair_ID[i] <- pair_id
    sorted_data$Pair_ID[i + 1] <- pair_id

    pair_id <- pair_id + 1
  }
}

# Handle unpaired unit (if odd number of units)
if (n_unpaired > 0) {
  sorted_data$Treatment_Final[n_units] <- sample(c(0, 1), 1)
  sorted_data$Pair_ID[n_units] <- -1
}


# --- Validate matching quality -----------------------------------------------
# Calculate within-pair differences in g_hat
paired_data <- sorted_data %>%
  filter(Pair_ID >= 0)

if (nrow(paired_data) > 0) {
  # Create pair-level dataset
  pairs_df <- paired_data %>%
    group_by(Pair_ID) %>%
    filter(n() == 2) %>%
    summarise(
      g_hat_diff = abs(diff(g_hat)),
      g_hat_mean = mean(g_hat),
      .groups = "drop"
    )

  # Calculate summary statistics
  mean_diff <- mean(pairs_df$g_hat_diff)
  median_diff <- median(pairs_df$g_hat_diff)
  max_diff <- max(pairs_df$g_hat_diff)
  min_diff <- min(pairs_df$g_hat_diff)

  # Calculate what the mean difference would be under random pairing
  random_diff_mean <- sd_g_hat * sqrt(2)
  reduction <- 100 * (1 - mean_diff / random_diff_mean)
}


# --- Merge with full data and save results -----------------------------------
# Merge back with original main data to get all variables
original_indices <- sorted_data$original_index
full_matched_data <- main_data[original_indices, ]

# Add the matching variables
full_matched_data$Treatment <- sorted_data$Treatment_Final
full_matched_data$Pair_ID <- sorted_data$Pair_ID
full_matched_data$g_hat <- sorted_data$g_hat
full_matched_data$g_0 <- sorted_data$g_0
full_matched_data$g_1 <- sorted_data$g_1

# TODO: Specify output file name
# This file will contain the main sample with Treatment and Pair_ID assigned
output_file <- file.path(output_dir, "optimal_matched_main_sample.dta")
write_dta(full_matched_data, output_file)

# Count treated and control
n_treated <- sum(sorted_data$Treatment_Final == 1)
n_control <- sum(sorted_data$Treatment_Final == 0)


# --- Print results -----------------------------------------------------------
cat("\n", strrep("=", 80), "\n", sep = "")
cat("OPTIMAL MATCHED-PAIRS RANDOMIZATION FROM PILOT STUDY\n")
cat(strrep("=", 80), "\n", sep = "")

cat("\nPILOT REGRESSION RESULTS\n")
cat(strrep("-", 80), "\n", sep = "")
cat(sprintf("  R-squared:            %.4f\n", r2_pilot))
cat(sprintf("  Adj. R-squared:       %.4f\n", r2_adj_pilot))
cat(sprintf("  Treatment coef:       %7.4f\n", coef_treatment))
cat(sprintf("  Treatment p-value:    %7.4f\n", pval_treatment))

cat("\nESTIMATED g_hat STATISTICS\n")
cat(strrep("-", 80), "\n", sep = "")
cat(sprintf("  Sample size:          %d\n", n_complete))
cat(sprintf("  Mean:                 %7.4f\n", mean_g_hat))
cat(sprintf("  Std:                  %7.4f\n", sd_g_hat))
cat(sprintf("  Min:                  %7.4f\n", min_g_hat))
cat(sprintf("  Max:                  %7.4f\n", max_g_hat))

cat("\nMATCHED PAIRS STATISTICS\n")
cat(strrep("-", 80), "\n", sep = "")
cat(sprintf("  Total units:          %d\n", n_units))
cat(sprintf("  Matched pairs:        %d\n", n_pairs))
cat(sprintf("  Unpaired units:       %d\n", n_unpaired))
cat(sprintf("  Pairing rate:         %.1f%%\n", 100 * n_pairs * 2 / n_units))

if (nrow(paired_data) > 0) {
  cat("\nWITHIN-PAIR g_hat DIFFERENCES\n")
  cat(strrep("-", 80), "\n", sep = "")
  cat(sprintf("  Mean:                 %7.6f\n", mean_diff))
  cat(sprintf("  Median:               %7.6f\n", median_diff))
  cat(sprintf("  Max:                  %7.6f\n", max_diff))
  cat(sprintf("  Min:                  %7.6f\n", min_diff))

  cat("\nCOMPARISON TO RANDOM PAIRING\n")
  cat(strrep("-", 80), "\n", sep = "")
  cat(sprintf("  Expected diff (random):  %7.6f\n", random_diff_mean))
  cat(sprintf("  Achieved diff (optimal): %7.6f\n", mean_diff))
  cat(sprintf("  Reduction:               %7.1f%%\n", reduction))
}

cat("\nTREATMENT ASSIGNMENT\n")
cat(strrep("-", 80), "\n", sep = "")
cat(sprintf("  Treated:              %d (%.1f%%)\n", n_treated, 100 * n_treated / n_units))
cat(sprintf("  Control:              %d (%.1f%%)\n", n_control, 100 * n_control / n_units))

cat("\n", strrep("=", 80), "\n", sep = "")
cat("OPTIMAL MATCHING COMPLETE\n")
cat(strrep("=", 80), "\n", sep = "")
cat("\nSaved to:", output_file, "\n\n")

# =============================================================================
# END OF OPTIMAL MATCHING SCRIPT
# =============================================================================

File: Ch. 06/Matched Pairs (Pilot)/code/optimal_matching_from_pilot.py
# =============================================================================
# Optimal Matched-Pairs Randomization Using Pilot Study
# =============================================================================
# Implements optimal matched-pairs randomization based on Bai (2022) framework
# for minimizing mean-squared error. Uses pilot study regression to estimate
# expected outcomes g_i = E[Y_i(1) + Y_i(0) | X_i] for main sample, then creates
# matched pairs by sorting on g_i and pairing adjacent units.
#
# Steps:
#   1. Load pilot data (with Treatment and Outcome from previous randomization)
#   2. Run regression on pilot to estimate relationship between covariates and outcomes
#   3. For main sample: predict outcomes under D=0 and D=1
#   4. Calculate g_hat = g_0 + g_1 for each unit
#   5. Sort main sample by g_hat
#   6. Create matched pairs from sorted data (pair adjacent units)
#   7. Within each pair, randomly assign treatment
#
# Section 6.3.5.1: Efficient Matching Minimizing Mean-Squared Error

from pathlib import Path
import numpy as np
import pandas as pd
import statsmodels.api as sm


# --- Setup -------------------------------------------------------------------
# Automatically resolve paths relative to this script's location.
SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR.parent / "data"
OUTPUT_DIR = SCRIPT_DIR.parent / "output"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


# --- Parameters --------------------------------------------------------------
# TODO: Set random seed for reproducibility
RANDOM_SEED = 42

# TODO: Specify control variables to use in g_i estimation
# These should be the covariates available in your dataset
# IMPORTANT: Must match the control variables used in the pilot regression
CONTROL_VARS = ['female', 'race_w', 'birthweight', 'std_ncog_pre', 'year']


# --- Initialize random state -------------------------------------------------
np.random.seed(RANDOM_SEED)


# --- Load pilot data and estimate g_i model ---------------------------------
# TODO: Specify the pilot data file
# This file should contain:
#   - Treatment variable (0/1 or binary)
#   - Outcome variable (continuous)
#   - All control variables specified in CONTROL_VARS
pilot_file = OUTPUT_DIR / 'pilot_sample_with_treatment_and_outcome.dta'
pilot_data = pd.read_stata(pilot_file)

# Prepare regression data (drop missing values in controls)
reg_vars = ['Outcome', 'Treatment'] + CONTROL_VARS
pilot_reg = pilot_data[reg_vars].dropna().copy()

# TODO: Modify regression specification if desired
# You can add interaction effects, polynomials, or other transformations
# Example: X['interaction'] = X['female'] * X['race_w']
# Example: X['birthweight_sq'] = X['birthweight'] ** 2
# Run regression: Outcome ~ Treatment + Controls
X = pilot_reg[['Treatment'] + CONTROL_VARS]
X = sm.add_constant(X)
y = pilot_reg['Outcome']

pilot_model = sm.OLS(y, X).fit()


# --- Load main sample and estimate g_0 and g_1 ------------------------------
# TODO: Specify the main sample data file
# This file should contain:
#   - All control variables specified in CONTROL_VARS
#   - NO Treatment variable (will be assigned by this script)
main_file = OUTPUT_DIR / 'main_sample.dta'
main_data = pd.read_stata(main_file)

# Keep only complete cases for prediction
main_complete = main_data[CONTROL_VARS].dropna().copy()
main_complete['original_index'] = main_complete.index

# TODO: If you modified the regression specification above,
# you must apply the SAME transformations here for prediction
# Example: If you added X['interaction'] = X['female'] * X['race_w'] above,
#          you must also add main_complete_d0['interaction'] = ... and main_complete_d1['interaction'] = ... below
# Predict under D=0 (Treatment=0)
main_complete_d0 = main_complete[CONTROL_VARS].copy()
main_complete_d0['Treatment'] = 0
X_d0 = sm.add_constant(main_complete_d0[['Treatment'] + CONTROL_VARS], has_constant='add')
g_0 = pilot_model.predict(X_d0)

# Predict under D=1 (Treatment=1)
main_complete_d1 = main_complete[CONTROL_VARS].copy()
main_complete_d1['Treatment'] = 1
X_d1 = sm.add_constant(main_complete_d1[['Treatment'] + CONTROL_VARS], has_constant='add')
g_1 = pilot_model.predict(X_d1)

# Add predictions to main_complete
main_complete['g_0'] = g_0
main_complete['g_1'] = g_1

# Calculate g_hat = g_0 + g_1
main_complete['g_hat'] = main_complete['g_0'] + main_complete['g_1']


# --- Sort by g_hat and create matched pairs ---------------------------------
# Sort by g_hat (increasing order)
sorted_data = main_complete.sort_values('g_hat').reset_index(drop=True)

# Create pairs from adjacent units in sorted order
n_units = len(sorted_data)
n_pairs = n_units // 2
n_unpaired = n_units % 2

# Initialize treatment and pair ID columns
sorted_data['Treatment_Final'] = -1
sorted_data['Pair_ID'] = -1

pair_id = 0

# Create pairs from adjacent units
for i in range(0, n_pairs * 2, 2):
    if i + 1 < len(sorted_data):
        # Randomly assign treatment within pair
        if np.random.random() < 0.5:
            sorted_data.loc[i, 'Treatment_Final'] = 1
            sorted_data.loc[i+1, 'Treatment_Final'] = 0
        else:
            sorted_data.loc[i, 'Treatment_Final'] = 0
            sorted_data.loc[i+1, 'Treatment_Final'] = 1

        # Assign same pair ID
        sorted_data.loc[i, 'Pair_ID'] = pair_id
        sorted_data.loc[i+1, 'Pair_ID'] = pair_id

        pair_id += 1

# Handle unpaired unit (if odd number of units)
if n_unpaired > 0:
    sorted_data.loc[len(sorted_data)-1, 'Treatment_Final'] = np.random.randint(0, 2)
    sorted_data.loc[len(sorted_data)-1, 'Pair_ID'] = -1


# --- Validate matching quality -----------------------------------------------
# Calculate within-pair differences in g_hat
paired_data = sorted_data[sorted_data['Pair_ID'] >= 0].copy()

if len(paired_data) > 0:
    pairs_list = []
    for pid in paired_data['Pair_ID'].unique():
        pair = paired_data[paired_data['Pair_ID'] == pid]
        if len(pair) == 2:
            g_vals = pair['g_hat'].values
            g_diff = abs(g_vals[0] - g_vals[1])
            g_mean = (g_vals[0] + g_vals[1]) / 2
            pairs_list.append({
                'Pair_ID': pid,
                'g_hat_diff': g_diff,
                'g_hat_mean': g_mean
            })

    pairs_df = pd.DataFrame(pairs_list)

    # Calculate what the mean difference would be under random pairing
    all_g_hat = sorted_data['g_hat'].values
    random_diff_mean = np.std(all_g_hat) * np.sqrt(2)
    reduction = 100*(1 - pairs_df['g_hat_diff'].mean()/random_diff_mean)


# --- Merge with full data and save results -----------------------------------
# Merge back with original main data to get all variables
original_indices = sorted_data['original_index'].astype(int).values
full_matched_data = main_data.iloc[original_indices].copy()

# Add the matching variables
full_matched_data['Treatment'] = sorted_data['Treatment_Final'].values
full_matched_data['Pair_ID'] = sorted_data['Pair_ID'].values
full_matched_data['g_hat'] = sorted_data['g_hat'].values
full_matched_data['g_0'] = sorted_data['g_0'].values
full_matched_data['g_1'] = sorted_data['g_1'].values

# TODO: Specify output file name
# This file will contain the main sample with Treatment and Pair_ID assigned
output_file = OUTPUT_DIR / 'optimal_matched_main_sample.dta'
full_matched_data.to_stata(output_file, write_index=False)


# --- Print results -----------------------------------------------------------
print("\n" + "=" * 80)
print("OPTIMAL MATCHED-PAIRS RANDOMIZATION FROM PILOT STUDY")
print("=" * 80)

print("\nPILOT REGRESSION RESULTS")
print("-" * 80)
print(f"  R-squared:            {pilot_model.rsquared:.4f}")
print(f"  Adj. R-squared:       {pilot_model.rsquared_adj:.4f}")
print(f"  Treatment coef:       {pilot_model.params['Treatment']:>7.4f}")
print(f"  Treatment p-value:    {pilot_model.pvalues['Treatment']:>7.4f}")

print("\nESTIMATED g_hat STATISTICS")
print("-" * 80)
print(f"  Sample size:          {len(main_complete)}")
print(f"  Mean:                 {main_complete['g_hat'].mean():>7.4f}")
print(f"  Std:                  {main_complete['g_hat'].std():>7.4f}")
print(f"  Min:                  {main_complete['g_hat'].min():>7.4f}")
print(f"  Max:                  {main_complete['g_hat'].max():>7.4f}")

print("\nMATCHED PAIRS STATISTICS")
print("-" * 80)
print(f"  Total units:          {n_units}")
print(f"  Matched pairs:        {n_pairs}")
print(f"  Unpaired units:       {n_unpaired}")
print(f"  Pairing rate:         {100*n_pairs*2/n_units:.1f}%")

if len(paired_data) > 0:
    print("\nWITHIN-PAIR g_hat DIFFERENCES")
    print("-" * 80)
    print(f"  Mean:                 {pairs_df['g_hat_diff'].mean():>7.6f}")
    print(f"  Median:               {pairs_df['g_hat_diff'].median():>7.6f}")
    print(f"  Max:                  {pairs_df['g_hat_diff'].max():>7.6f}")
    print(f"  Min:                  {pairs_df['g_hat_diff'].min():>7.6f}")

    print("\nCOMPARISON TO RANDOM PAIRING")
    print("-" * 80)
    print(f"  Expected diff (random):  {random_diff_mean:>7.6f}")
    print(f"  Achieved diff (optimal): {pairs_df['g_hat_diff'].mean():>7.6f}")
    print(f"  Reduction:               {reduction:>7.1f}%")

n_treated = (sorted_data['Treatment_Final'] == 1).sum()
n_control = (sorted_data['Treatment_Final'] == 0).sum()
print("\nTREATMENT ASSIGNMENT")
print("-" * 80)
print(f"  Treated:              {n_treated} ({100*n_treated/len(sorted_data):.1f}%)")
print(f"  Control:              {n_control} ({100*n_control/len(sorted_data):.1f}%)")

print("\n" + "=" * 80)
print("OPTIMAL MATCHING COMPLETE")
print("=" * 80)
print(f"\n✓ Saved to: {output_file}\n")

# =============================================================================
# END OF OPTIMAL MATCHING SCRIPT
# =============================================================================
